src/mpid/ch4/netmod/ofi/ofi_win.c

```diff
@@ -189,56 +192,73 @@ static int win_allgather(MPIR_Win * win, void *base, int disp_unit)
         MPIDI_OFI_WIN(win).mr_key = 0;
     }
 
-    /* It is unclear whether we can register NULL and what that means even in the case
-     * we can. So until there is clear documentation, we take the conservative measure
-     * and do not register a NULL base. */
-
-    /* Implication for dynamic window: since we are not registering mr here, we currently
-     * always fall-back to active messages for dynamic window. In any way, dynamic window
-     * probably will need per-attachment registration and a mechanism for dynamic
-     * synchronization to work with direct RMA. That is difficult without extra sematics or
-     * hints, unfortunately. Until then, we fall back. */
-
-    if (base) {
-        MPIDI_OFI_CALL(fi_mr_reg(MPIDI_OFI_global.ctx[0].domain,        /* In:  Domain Object */
-                                 base,  /* In:  Lower memory address */
-                                 win->size,     /* In:  Length              */
-                                 FI_REMOTE_READ | FI_REMOTE_WRITE,      /* In:  Expose MR for read  */
-                                 0ULL,  /* In:  offset(not used)    */
-                                 MPIDI_OFI_WIN(win).mr_key,     /* In:  requested key       */
-                                 0ULL,  /* In:  flags               */
-                                 &MPIDI_OFI_WIN(win).mr,        /* Out: memregion object    */
-                                 NULL), mr_reg);        /* In:  context             */
+    /* fi_mr_reg may fail for legitimate reason, for example, with dynamic window, we
+     * are trying to register MPI_BOTTOM, some providers can legitimately reject.
+     *
+     * When fi_mr_reg fails, we should ensure always fallback to active message.
+     */
+
+    if (win->size > 0) {
+        int rc = fi_mr_reg(MPIDI_OFI_global.ctx[0].domain, base, win->size,
+                           FI_REMOTE_READ | FI_REMOTE_WRITE, 0ULL /* offset */ ,
+                           MPIDI_OFI_WIN(win).mr_key, 0ULL /* flags */ ,
+                           &MPIDI_OFI_WIN(win).mr, NULL /* context */);
+        if (rc < 0) {
+            /* Failed for whatever reason. Set mr_key to 0, which will enforce the fallback */
+            MPIDI_OFI_WIN(win).mr_key = MPIDI_OFI_INVALID_MR_KEY;
```

Comment:
Can we fallback to AM only when the `rc` is a specific error? For any unexpected error, we should just throw the error out.

Reply:
> When `FI_MR_VIRT_ADDR ` is set (basic MR), I think we can skip `fi_mr_reg` because we know it always fails.

I think we don't know it always fails. From the discussion we had yesterday, `FI_MR_VIRT_ADDR` just says about the semantics used if libfabric calls such as `fi_readmsg` and `fi_writemsg`. It has nothing to do with how or whether it needs memory registration.

Reply:
> Can we fallback to AM only when the `rc` is a specific error? For any unexpected error, we should just throw the error out.

Currently, all errors in `fi_mr_reg` will force us to fallback to AM. If we want to allow certain error, then we need know and put code specifically to handle that error. What specific error code do you have in mind?